pkg/controller/controller.go

```diff
@@ -418,75 +433,145 @@ func (c *Controller) runAction(ctx context.Context, as *crv1alpha1.ActionSet, aI
 			} else {
 				rf = func(ras *crv1alpha1.ActionSet) error {
 					ras.Status.Actions[aIDX].Phases[i].State = crv1alpha1.StateComplete
+					// this updates the phase output in the actionset status
 					ras.Status.Actions[aIDX].Phases[i].Output = output
 					return nil
 				}
 			}
+
 			if rErr := reconcile.ActionSet(ctx, c.crClient.CrV1alpha1(), ns, name, rf); rErr != nil {
 				reason := fmt.Sprintf("ActionSetFailed Action: %s", as.Spec.Actions[aIDX].Name)
 				msg := fmt.Sprintf("Failed to update phase: %#v:", as.Status.Actions[aIDX].Phases[i])
 				c.logAndErrorEvent(ctx, msg, reason, rErr, as, bp)
+				c.executeDeferPhase(ctx, deferPhase, tp, bp, action.Name, aIDX, as, rErr)
 				return nil
 			}
+
 			if err != nil {
 				reason := fmt.Sprintf("ActionSetFailed Action: %s", as.Spec.Actions[aIDX].Name)
 				if msg == "" {
 					msg = fmt.Sprintf("Failed to execute phase: %#v:", as.Status.Actions[aIDX].Phases[i])
 				}
 				c.logAndErrorEvent(ctx, msg, reason, err, as, bp)
+				c.executeDeferPhase(ctx, deferPhase, tp, bp, action.Name, aIDX, as, err)
```

Comment:
Instead of calling this in multiple places, can we use a Go `defer` like:

```diff
diff --git a/pkg/controller/controller.go b/pkg/controller/controller.go
index 21d0375c..b7660727 100644
--- a/pkg/controller/controller.go
+++ b/pkg/controller/controller.go
@@ -409,6 +409,10 @@ func (c *Controller) runAction(ctx context.Context, as *crv1alpha1.ActionSet, aI
 	c.actionSetTombMap.Store(as.Name, t)
 	ctx = field.Context(ctx, consts.ActionsetNameKey, as.GetName())
 	t.Go(func() error {
+		var coreErr error
+		defer func() {
+			c.executeDeferPhase(ctx, deferPhase, tp, bp, action.Name, aIDX, as, coreErr)
+		}()
 		for i, p := range phases {
 			ctx = field.Context(ctx, consts.PhaseNameKey, p.Name())
 			c.logAndSuccessEvent(ctx, fmt.Sprintf("Executing phase %s", p.Name()), "Started Phase", as)
@@ -443,7 +447,7 @@ func (c *Controller) runAction(ctx context.Context, as *crv1alpha1.ActionSet, aI
 				reason := fmt.Sprintf("ActionSetFailed Action: %s", as.Spec.Actions[aIDX].Name)
 				msg := fmt.Sprintf("Failed to update phase: %#v:", as.Status.Actions[aIDX].Phases[i])
 				c.logAndErrorEvent(ctx, msg, reason, rErr, as, bp)
-				c.executeDeferPhase(ctx, deferPhase, tp, bp, action.Name, aIDX, as, rErr)
+				coreErr = rErr
 				return nil
 			}
 
@@ -453,13 +457,12 @@ func (c *Controller) runAction(ctx context.Context, as *crv1alpha1.ActionSet, aI
 					msg = fmt.Sprintf("Failed to execute phase: %#v:", as.Status.Actions[aIDX].Phases[i])
 				}
 				c.logAndErrorEvent(ctx, msg, reason, err, as, bp)
-				c.executeDeferPhase(ctx, deferPhase, tp, bp, action.Name, aIDX, as, err)
+				coreErr = err
 				return nil
 			}
 			param.UpdatePhaseParams(ctx, tp, p.Name(), output)
 			c.logAndSuccessEvent(ctx, fmt.Sprintf("Completed phase %s", p.Name()), "Ended Phase", as)
 		}
-		c.executeDeferPhase(ctx, deferPhase, tp, bp, action.Name, aIDX, as, nil)
 		return nil
 	})
 	return nil
```